tools/rofiTool/common.cpp

```diff
@@ -0,0 +1,41 @@
+#include "common.hpp"
+
+#include <fstream>
+#include <stdexcept>
+
+#include <atoms/result.hpp>
+#include <configuration/rofiworld.hpp>
+#include <configuration/serialization.hpp>
+#include <configuration/universalModule.hpp>
+
+
+atoms::Result< rofi::configuration::RofiWorld > parseRofiWorld( const std::string & inputFile )
+{
+    using namespace rofi::configuration;
+
+    auto cfgFile = std::ifstream( inputFile );
+    if ( !cfgFile.is_open() ) {
+        return atoms::result_error( "Cannot open file '" + inputFile + "'" );
+    }
+
+    try {
+        if ( inputFile.ends_with( ".json" ) ) {
+            auto json = nlohmann::json::parse( cfgFile );
+            return atoms::result_value( serialization::fromJSON( json ) );
+        }
+
+        return atoms::result_value( readOldConfigurationFormat( cfgFile ) );
+    } catch ( const std::exception & e ) {
+        return atoms::result_error< std::string >( e.what() );
+    }
+}
+
+void affixRofiWorld( rofi::configuration::RofiWorld & world )
+{
+    using namespace rofi::configuration;
+    auto modules = world.modules();
+    assert( !modules.empty() );
+    const auto & firstModule = modules.begin()->module;
+    assert( !firstModule->bodies().empty() );
+    connect< RigidJoint >( firstModule->bodies().front(), {}, matrices::identity );
```

Comment:
Well, no, because it is used in a translation (only x,y,z values are used). The w value is normally used for scaling - but the implementation here just ignores it.

Reply:
So should the last element of a reference point always be 0? I am working with reference points in the "points" command of rofi-tool, where I am translating them to a new coordinate system; should they always be in the (x,y,z,0) format?

Reply:
When you look into the code, the `refPoint` is only ever accessed at `rofiworld.cpp:161` in the form `translate( j.refPoint )`. And since `translate` ignores the `w` part, it's just irrelenvant what you store there.

If you want to use a point for multiplicating with a matrix, you have to make sure it has the correct `w` (=1)... but you also can't rely that the `refPoint` in configuration will have the `w` value set.
(Sidenote: You can get the position from a matrix by taking the last column.)